design/proof-network.md

```diff
@@ -0,0 +1,257 @@
+
+### Missing proofs ###
+
+Validators monitor the on-chain marketplace to check which slots require a
+storage proof to be submitted and what the public *inputs* for the proof are.
+For each required proof they check at the end of its *period* whether that proof
+was submitted and correct. If they did not receive a correct proof then they
+will add a `MissingProof(slot id, period, inputs)` transaction to the next
+block.
+
+The `MissingProof` transactions are sequenced by the consensus algorithm. They
+are then evaluated by each validator. They will note the first `MissingProof`
+transaction that correctly notices a missing proof, and allow the sender of that
+first transaction to go on-chain to mark the proof as missing.
+
+The validator that sent the first `MissingProof` transaction for a missing proof
+can now request BLS signatures from the other validators to enact on-chain
+slashing of the storage provider for missing a proof. When the validator
+receives enough signatures to represent > 2/3 stake it can combine these
+signatures into a single combined BLS signature. The validator can then submit
+*slot id*, *period*, *inputs* and the combined signature to the marketplace.
+
+The marketplace will then verify the correctness of *inputs* for the *slot id*
+and *period*, and checks that the combined signature is representative for > 2/3
+stake. If these conditions are met, it will then slash the storage provider
+collateral and reward the validator.
+
+### Faulty proofs ###
+
+The storage proofs that a storage provider submits can be faulty for a number of
+reasons:
+
+1. The zero knowledge *proof* is incorrect
+2. The submitted *period* is not the current time period
+3. The public *inputs* to the proof do not match the values from the on-chain
+   marketplace
+
+Faults 1 and 2 are caught by the validators. Validators check the zero-knowledge
+*proof*, and that the *period* is the current time period when evaluating a
+`StorageProof` transaction. If these are incorrect then they ignore the
+transaction, effectively treating the proof as missing. Validators now go
+through the same flow that we described in the previous section.
+
+Fault 3 is caught by the validators and the on-chain marketplace. Validators
+will look for a correct `StorageProof` transaction that has the same *inputs* as
+specified by the marketplace. If it doesn't find it because the storage provider
+submitted a `StorageProof` transaction with a different value for *inputs*, then
+it will treat the proof as missing, and go through the same flow as in the
+previous section.
+
+Consensus
+---------
+
+Our design depends on a consensus algorithm between validators. This can be any
+byzantine-fault-tolerant consensus algorithm for sequencing transactions. The
+[Mysticeti][6] algorithm seems to be particularly suited because it is highly
+performant.
+
+The consensus algorithm is only used to sequence transactions. Evaluation of the
```

Comment:
Out of curiosity: why does sequencing transactions matter?

Reply:
I guess, maybe to match the right/current period? If a transaction is added too late will it miss the period?

Reply:
Sequencing is not strictly necessary for storage proofs, but for anything else it becomes necessary.

See my reply here: https://github.com/codex-storage/codex-research/pull/197/files#r1820368545